robot/impl/optional_dependencies_test.go

```diff
@@ -0,0 +1,346 @@
+	ocName := generic.Named("oc")
+	resource.Register(
+		generic.API,
+		optionalChildModel,
+		resource.Registration[*optionalChild, *optionalChildConfig]{
+			Constructor: newOptionalChild,
+		})
+	defer resource.Deregister(generic.API, optionalChildModel)
+
+	// Reconfigure the robot to have an optional child component, its required motor, and no
+	// optional motor.
+	cfg := config.Config{
+		Components: []resource.Config{
+			{
+				Name:  ocName.Name,
+				API:   generic.API,
+				Model: optionalChildModel,
+				ConvertedAttributes: &optionalChildConfig{
+					RequiredMotor: "m",
+					OptionalMotor: "m1",
+				},
+			},
+			{
+				Name:                "m",
+				API:                 motor.API,
+				Model:               fake.Model,
+				ConvertedAttributes: &fake.Config{},
+			},
+		},
+	}
+	// Ensure here and for all configs below before `Reconfigure`ing to make sure optional
+	// dependencies are calculated (`ImplicitOptionalDependsOn` is filled in).
+	test.That(t, cfg.Ensure(false, logger), test.ShouldBeNil)
+	lr.Reconfigure(ctx, &cfg)
+
+	{ // Assertions
+		// Assert that the optional child component built successfully (optional dependency on
+		// non-existent 'm1' did not cause a failure to build).
+		ocRes, err := lr.ResourceByName(ocName)
+		test.That(t, err, test.ShouldBeNil)
+
+		// Assert that the optional child reconfigured and logged its inability to get 'm1'
+		// from dependencies _twice_. The first of both is from construction (invokes
+		// `Reconfigure`) of the resource, and the second of both is from reconfiguring of the
+		// resource due to an unconditional call to `updateWeakAndOptionalDependents` directly
+		// after `completeConfig`.
+		oc, err := resource.AsType[*optionalChild](ocRes)
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, oc.reconfigCount, test.ShouldEqual, 2)
+		msgNum := logs.FilterMessageSnippet("could not get optional motor").Len()
+		test.That(t, msgNum, test.ShouldEqual, 2)
+
+		// Assert that, on the component itself, `requiredMotor` is now set, but `optionalMotor`
+		// is not.
+		test.That(t, oc.requiredMotor, test.ShouldNotBeNil)
+		test.That(t, oc.optionalMotor, test.ShouldBeNil)
+	}
+
+	// Reconfigure the robot to have the optionally-depended-upon motor 'm1'.
+	cfg = config.Config{
+		Components: []resource.Config{
+			{
+				Name:  ocName.Name,
+				API:   generic.API,
+				Model: optionalChildModel,
+				ConvertedAttributes: &optionalChildConfig{
+					RequiredMotor: "m",
+					OptionalMotor: "m1",
+				},
+			},
+			{
+				Name:                "m",
+				API:                 motor.API,
+				Model:               fake.Model,
+				ConvertedAttributes: &fake.Config{},
+			},
+			{
+				Name:                "m1",
+				API:                 motor.API,
+				Model:               fake.Model,
+				ConvertedAttributes: &fake.Config{},
+			},
+		},
+	}
+	test.That(t, cfg.Ensure(false, logger), test.ShouldBeNil)
+	lr.Reconfigure(ctx, &cfg)
+
+	{ // Assertions
```

Comment:
never seen braces by themselves before

Reply:
Yeah it's a pattern I picked up from a Go driver engineer at MongoDB. Functionally, all it does is create a new scope (e.g. `err` has to be redefined with `:=` instead of reassigned with `=`). Stylistically, I find it helpful to group small, logical sections of code that are not otherwise delimited with a conditional or function into their sections.